Add AutoDiscovery for Kubernetes

J12934 opened the pull request:
## Description

This PR adds a "AutoDiscovery" feature for workloads running inside kubernetes.
Currently only supports http / https types of workloads by "auto-discovering" kubernetes services and starting scans against them.

For more info check out the readme of the AutoDiscovery subfolder: https://github.com/secureCodeBox/secureCodeBox/tree/auto-discovery/auto-discovery/kubernetes

### Todos

- [x] License headers
- [x] Resources limits configuration via helm chart

fuhrmeistery commented on .github/workflows/ci.yaml:
```diff
@@ -174,6 +174,59 @@ jobs:
+        with:
+          go-version: ${{ env.GO_VERSION }}
+
+      - name: Lint Go Code
+        run: |
+          cd auto-discovery/kubernetes/
+          go fmt ./...
+          go vet ./...
```
```suggestion
      - name: Lint Go Code
        working-directory: ./auto-discovery/kubernetes
        run: |
          go fmt ./...
          go vet ./...

fuhrmeistery commented on .github/workflows/ci.yaml:
```diff
@@ -174,6 +174,59 @@ jobs:
+          go fmt ./...
+          go vet ./...
+
+      - name: Test
+        run: |
+          cd auto-discovery/kubernetes/
+          make test
```
```suggestion
      - name: Test
        working-directory: ./auto-discovery/kubernetes/
        run: make test
```

fuhrmeistery reviewed (commented):
Great Work!
I just have some questions and minor issues.
1. Why do we need a `kubernetes` directory?
2. I think it would be best to include the short SPDX Header instead of the Apache Header
3. Trying to run the tests locally with `make test` fails with

```bash
•STEP: tearing down the test environment

------------------------------
Failure [20.014 seconds]
[AfterSuite] AfterSuite
/home/yfuhrmeister/Projects/scb/secureCodeBox/auto-discovery/kubernetes/controllers/suite_test.go:119

  Unexpected error:
      <errors.aggregate | len:1, cap:1>: [
          {
              s: "timeout waiting for process kube-apiserver to stop",
          },
      ]
      timeout waiting for process kube-apiserver to stop
  occurred

  /home/yfuhrmeister/Projects/scb/secureCodeBox/auto-discovery/kubernetes/controllers/suite_test.go:122
```

J12934 commented on auto-discovery/kubernetes/Dockerfile:
```diff
@@ -0,0 +1,26 @@
+COPY pkg/ pkg/
+
+# Build
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
```
This is directly taken from the defaukt kubebuilder starting point, so not sure what the exact reasoning here is to doing it like this, I'd guess that this gives them more controll over the flags used for the go build but I'm not sure.

J12934 commented:
> Great Work!
> I just have some questions and minor issues.
> 
>     1. Why do we need a `kubernetes` directory?
> 
>     2. I think it would be best to include the short SPDX Header instead of the Apache Header
> 
>     3. Trying to run the tests locally with `make test` fails with
> 
> 
> ```shell
> •STEP: tearing down the test environment
> 
> ------------------------------
> Failure [20.014 seconds]
> [AfterSuite] AfterSuite
> /home/yfuhrmeister/Projects/scb/secureCodeBox/auto-discovery/kubernetes/controllers/suite_test.go:119
> 
>   Unexpected error:
>       <errors.aggregate | len:1, cap:1>: [
>           {
>               s: "timeout waiting for process kube-apiserver to stop",
>           },
>       ]
>       timeout waiting for process kube-apiserver to stop
>   occurred
> 
>   /home/yfuhrmeister/Projects/scb/secureCodeBox/auto-discovery/kubernetes/controllers/suite_test.go:122
> ```

@fuhrmeistery 

1. The kubernetes directory is so that we can add additional "auto-discoveries" for different cloud / runtime environments in the future, not just kubernetes.
2. Yup, good point the existing headers are automatically setup by kubebuilder didn't setup the spdx header up yet...
3. Not sure whats going on there, seems to work okay for my machine and the CI, looks like this is only failing to tear down the api server, is your linux distro configured in a unusual way which might interfer with that?

fuhrmeistery commented:
> 
> @fuhrmeistery
> 
>     1. The kubernetes directory is so that we can add additional "auto-discoveries" for different cloud / runtime environments in the future, not just kubernetes. This is mentioned in the readme in the `auto-discovery` dir.
> 
>     2. Yup, good point the existing headers are automatically setup by kubebuilder didn't setup the spdx header up yet...
> 
>     3. Not sure whats going on there, seems to work okay for my machine and the CI, looks like this is only failing to tear down the api server, is your linux distro configured in a unusual way which might interfer with that?
3. Yes, I think you are right. Tested on another machine and everything was fine o.O

J12934 commented:
Mh interesting, hopefully that's actually related to your system and not that the test suite is flaky 😅

rseedorff commented:
Resource Configuration would be great

J12934 commented on auto-discovery/kubernetes/README.md:
```diff
@@ -0,0 +1,132 @@
+The secureCodeBox _auto-discovery_ is running on kubernetes (K8S) and is an optional component of the complete secureCodeBox stack.
+The Kubernetes Auto-Discovery needs to be deployed along side the secureCodeBox Operator. It monitors security relevant resources inside a K8S environment and automatically create scans to continuously monitor security aspects of the resources.
+
+The operator will automatically detect these new resources (services) and start secureCodeBox _scans_ for them:
```
That not acutally true. Its finding the resources (k8s services) in the kubernetes api, it doesn't perform any port scans by default.

J12934 commented on auto-discovery/kubernetes/demo/juice-shop.yaml:
```diff
@@ -0,0 +1,81 @@
+#   values:
+#     defectdojo:
+#       auth:
+#         apiKey: e2ed466e0f85127d08879c178ec8ec8df3a15452
```
Yup totally, is not removed, apiKey was never acutally valid. (at least not at the point in time where it was commited.)

J12934 commented on auto-discovery/kubernetes/api/v1/autodiscoveryconfig_types.go:
```diff
@@ -0,0 +1,68 @@
+	Name string `json:"name"`
+}
+
+// HookState Describes the State of a Hook on a Scan
```
Nope, good point 👍

J12934 commented on auto-discovery/kubernetes/controllers/service_scan_controller.go:
```diff
@@ -0,0 +1,464 @@
+			"auto-discovery.securecodebox.io/target-port":    fmt.Sprintf("%d", host.Port),
+		}
+		for containerName, podDigest := range podDigests {
+			// The map should only contain one entry at this point. As the reconciler breaks if the services points at multiple version before
```
Cleared the comment up up bit.

Basically this was meant to explain that at this point in the reconciller we are sure that we only ever have one cannonical digest for a container name.

E.g. you have a juice-shop k8s deployment which is currently being upgraded from 12.8.0 to 12.8.1. During the upgrade service might be temporrily backed by different versions of the image.
We dont want to start the scans while the upgrade is still underway, as we only want to ensure that we only scan the new version as we assume that we have already scanned the old version.

This comment is meant to explain that the next lines would not make sense without it as they assign a unique digest per containerName.

J12934 commented:
- I fixed some typos in the helm-docs templates this has exploded as this has changed the doc files in every component. PR should still be reviewable by just checking the changes `auto-discovery/` and `.helm-docs/` folder.
- I changed the helm-docs action to check the checksum of the downloaded binary and mv'ed it to the `/usr/local/bin` so that we don't have to juggle with the correct relative path every time

fuhrmeistery reviewed (approved):
Looks great to me :+1:
